### DISCUSSION TITLE
Code gets blocked on double use

### DISCUSSION THREAD

**Discussion by HilkopterBob** [zauberzeug/nicegui#704] [Q&A]
### Question

Hey i have following problem:

I implemented search as you type from your example, added a card container for results with checboxes, a submit button to submit checked results and a container card to hold returned results with a clear button. So pretty simple. But if i use this code 2x for one page in two different page segments it does only work for the last element. 

---

My code:

parent element for the searchbars:
```py
with ui.column().classes("w-full self-center").bind_visibility_from(host_groups, "value"):
    with ui.row().classes("w-full asolute-center self-center"):
        ui.button("Submit", on_click= callfunc_place_submitted).classes("self-center")
        await sayt(r"lookup.csv")        # sayt = Search as you type
        global searchresults_column_hostgroups
    searchresults_column_hostgroups = ui.column().classes("w-full items-center")
        
with ui.column().classes("w-full self-center").bind_visibility_from(host_individual, "value"):
    with ui.row().classes("w-full asolute-center self-center"):
        ui.button("Submit", on_click=callfunc_place_submitted).classes("center self-center")
        await sayt(r"lookup.csv")        # sayt = Search as you type
        global searchresults_column_hostindividual
    searchresults_column_hostindividual = ui.column().classes("w-full items-center")
```

sayt:
```py
async def sayt (file):
    global checked_results
    global search_field
    global results
    global results_container
    global file_path
    file_path = file
    search_field = ui.input(on_change=search) \
        .props('input-class="ml-3"') \
        .classes('w-96 self-center transition-all')
    results_container = ui.card().classes('w-full items-center').bind_visibility_from(search_field, "value")
    return checked_results
```

callfuncttion to call all needed functions for placing and cleaning of results without parameters:
```py
async def callfunc_place_submitted():
    await place_submitted_search_values(searchresults_column_hostgroups)
    await clear_checked_results()
    ui.notify("submitted", type="info")
```

func for placing submitted results
```py
async def place_submitted_search_values(container):
    search_values = await get_checked_results()
    with container:
        for item in search_values:
            with ui.row().classes("w-full") as row:
                ui.label(text=item)
                ui.button("", on_click=lambda e, row=row: container.remove(row)  ).props("icon=delete")
```

search and place func:
```py
async def search(e: events.ValueChangeEventArguments) -> None:
    global running_query
    if running_query:
        running_query.cancel()
    results_container.clear()
    if e.value == "":
        await clear_checked_results()
    running_query = search_in_csv(e.value)
    response = await running_query
    for index, item in enumerate(response):
        if item == "":
            response.pop(index)
    if response != []:
        with results_container:  
            for item in response:  # iterate over the response data of the api
                with ui.row().classes('w-full items-center self-center') as row:
                    checkbox = ui.checkbox(text=item, on_change=check).classes('self-center')
    running_query = None
```

search in csv func
```py
async def search_in_csv(search_term):
    response = []
    with open(file_path, 'rt') as f:
        reader = csv.reader(f, delimiter=',')
        for _row in reader:
            if str(search_term) in str(_row):
                response.append(str(_row))
    return response
```

as you can see its quite a pile of code i needed to make it work, somewhat at least.

So after this code, do you know how it can be that it gets blocked every time i use it multiple times on one page?

**Comment by falkoschindler** [comment]
Sorry, but that's really quite some code and the code blocks you provided are not trivial to but together.
Can you, please, either boil it further down to a smaller example, or show the code as a whole? Thanks!

**Reply by HilkopterBob** [reply]
yes give me some time so i can talk to hr about how much code i can show. i will reply soon

**Comment by HilkopterBob** [comment]
this is the full code of my search-as-you-type component:

```py
"""search as you type Searchbar component
"""
import asyncio
from typing import Optional
import csv
from nicegui import events, ui


async def search_in_csv(search_term): ...

checked_results = []
running_query: Optional[asyncio.Task] = None

async def search(e: events.ValueChangeEventArguments) -> None:
    global running_query
    if running_query:
        running_query.cancel()
    results_container.clear()
    if e.value == "":
        await clear_checked_results()
    running_query = search_in_csv(e.value)
    response = await running_query
    for index, item in enumerate(response):
        if item == "":
            response.pop(index)
    if response != []:
        with results_container:  
            for item in response:  
                with ui.row().classes('w-full items-center self-center') as row:
                    checkbox = ui.checkbox(text=item, on_change=check).classes('self-center')
    running_query = None

async def sayt (file):
    global checked_results
    global search_field
    global results
    global results_container
    global file_path
    file_path = file
    search_field = ui.input(on_change=search) \
        .props('input-class="ml-3"') \
        .classes('w-96 self-center transition-all')
    results_container = ui.card().classes('w-full items-center').bind_visibility_from(search_field, "value")
    return checked_results

async def get_checked_results(): ...

async def check(event: events.ValueChangeEventArguments): ... 

async def clear_checked_results(): ...
    
```

---

@falkoschindler 
Its mainly the same as in your search as you type example from nicegui.io, besides some helper functions for making, retrieving and clearing of a list named "checked-results" in wich search returns are saved when the user chooses them (see checkbox in "`search`"). Another function i added does the searching in a csv.

My initial problem was, that i used the code two times in different parrent containers. I stll dont know how it can be blocked so its only functional in the last one its requested. But i solved it by making an exact copy and appending every function name with "_copy". wich works mostly. 

The new problem i have is, that the coe randomly stops working _fully_. even the exact copy stops working. The visible simptopms (no cli errors) are that in the result container, wich is returned, is nothing. its just an empty card. This happens every other time i restart the app or when it reloads itself (after saving a file). 

TL;DR: i stll cannot wrap my head aroud that buggy mess i wrote 😅

**Reply by falkoschindler** [reply]
Ok. It's a bit tough to help with just code fragments. I'd suggest you debug your code, simplify it, and remove any unrelated parts that don't cause the bug you're dealing with. Once you've got a minimal example or a clear problem statement, feel free to share it with us, and we'll be glad to help you out!
